{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "This notebook helps to prepare a training data for the CTD<sup>2</sup> Pancancer Drug Activity DREAM Challenge.\n",
    "\n",
    "Created by: Bence Szalai, Semmelweis University, Budapest, Hungary\n",
    "\n",
    "We will use the gene expression profiles from [LINCS-L1000 project](https://www.sciencedirect.com/science/article/pii/S0092867417313090). The compound metadata is coming from [The Drug Repurposing Hub](https://www.nature.com/articles/nm.4306). The analysis used here is part of [our previous manuscript](https://academic.oup.com/nar/article/47/19/10010/5573547?guestAccessKey=151831e1-0f78-4c70-854c-a471a28a5eac). We will also use the [cmapPy library](https://academic.oup.com/bioinformatics/article/35/8/1427/5094509). If you use this analyis in your work / following manuscript (after the challange publication embargo) please cite all of them appropriately. \n",
    "\n",
    "You have to download the data from Gene Expression Omnibus. We will use [GSE92742](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742) and [GSE70138](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE70138) datasets. You have to download from both datasets the following files:\n",
    "\n",
    "* Level5 gene expression profiles (GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx.gz and GSE70138_Broad_LINCS_Level5_COMPZ_n118050x12328_2017-03-06.gctx.gz respectively)\n",
    "* metadata from sig_info files (GSE92742_Broad_LINCS_sig_info.txt.gz and GSE70138_Broad_LINCS_sig_info_2017-03-06.txt.gz respectively)\n",
    "* gene info file (GSE92742_Broad_LINCS_gene_info.txt.gz, no need to download it from GSE70138, they are the same)\n",
    "\n",
    "Please download (and `gunzip`) these files, and put their content into the **data/GSE92742** and **data/GSE70138** folders of this project.\n",
    "\n",
    "You will also need compound metadata from [The Drug Repurposing Hub](https://clue.io/repurposing). You have to download the following files:\n",
    "* Drug information (updated 9/7/2018)\n",
    "* Sample information (updated 9/7/2018)\n",
    "\n",
    "Please download these files and put them to **data/repurposing**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#let's check files are in the correct directory\n",
    "import os\n",
    "files_needed_GSE92742=['GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx',\n",
    "               'GSE92742_Broad_LINCS_sig_info.txt','GSE92742_Broad_LINCS_gene_info.txt']\n",
    "files_needed_GSE70138=['GSE70138_Broad_LINCS_Level5_COMPZ_n118050x12328.gctx',\n",
    "                      'GSE70138_Broad_LINCS_sig_info.txt']\n",
    "files_needed_rep=['repurposing_drugs_20180907.txt','repurposing_samples_20180907.txt']\n",
    "files_GSE92742=os.listdir('../data/GSE92742/')\n",
    "files_GSE70138=os.listdir('../data/GSE70138/')\n",
    "files_rep=os.listdir('../data/repurposing/')\n",
    "for f in files_needed_GSE92742:\n",
    "    assert (f in files_GSE92742)\n",
    "for f in files_needed_GSE70138:\n",
    "    assert (f in files_GSE70138)\n",
    "for f in files_needed_rep:\n",
    "    assert (f in files_rep)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "You will need to install the [cmapPy library](https://clue.io/cmapPy/build.html#install)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/benceszalai/Library/Enthought/Canopy/edm/envs/Python3/lib/python3.5/site-packages/h5py/__init__.py:34: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  from ._conv import register_converters as _register_converters\n"
     ]
    }
   ],
   "source": [
    "# general data manipulations\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "\n",
    "#cmapPy\n",
    "from cmapPy.pandasGEXpress.parse import parse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "First we will get familiar with LINCS-L1000 data, and get shRNA related perturabtions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pert_id</th>\n",
       "      <th>pert_iname</th>\n",
       "      <th>pert_type</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>pert_dose</th>\n",
       "      <th>pert_dose_unit</th>\n",
       "      <th>pert_idose</th>\n",
       "      <th>pert_time</th>\n",
       "      <th>pert_time_unit</th>\n",
       "      <th>pert_itime</th>\n",
       "      <th>distil_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sig_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AML001_CD34_24H:A05</th>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>ctl_vehicle</td>\n",
       "      <td>CD34</td>\n",
       "      <td>0.1</td>\n",
       "      <td>%</td>\n",
       "      <td>0.1 %</td>\n",
       "      <td>24</td>\n",
       "      <td>h</td>\n",
       "      <td>24 h</td>\n",
       "      <td>AML001_CD34_24H_X1_F1B10:A05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AML001_CD34_24H:A06</th>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>ctl_vehicle</td>\n",
       "      <td>CD34</td>\n",
       "      <td>0.1</td>\n",
       "      <td>%</td>\n",
       "      <td>0.1 %</td>\n",
       "      <td>24</td>\n",
       "      <td>h</td>\n",
       "      <td>24 h</td>\n",
       "      <td>AML001_CD34_24H_X3_F1B10:A06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AML001_CD34_24H:B05</th>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>ctl_vehicle</td>\n",
       "      <td>CD34</td>\n",
       "      <td>0.1</td>\n",
       "      <td>%</td>\n",
       "      <td>0.1 %</td>\n",
       "      <td>24</td>\n",
       "      <td>h</td>\n",
       "      <td>24 h</td>\n",
       "      <td>AML001_CD34_24H_X1_F1B10:B05|AML001_CD34_24H_X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AML001_CD34_24H:B06</th>\n",
       "      <td>DMSO</td>\n",
       "      <td>DMSO</td>\n",
       "      <td>ctl_vehicle</td>\n",
       "      <td>CD34</td>\n",
       "      <td>0.1</td>\n",
       "      <td>%</td>\n",
       "      <td>0.1 %</td>\n",
       "      <td>24</td>\n",
       "      <td>h</td>\n",
       "      <td>24 h</td>\n",
       "      <td>AML001_CD34_24H_X3_F1B10:B06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AML001_CD34_24H:BRD-A03772856:0.37037</th>\n",
       "      <td>BRD-A03772856</td>\n",
       "      <td>BRD-A03772856</td>\n",
       "      <td>trt_cp</td>\n",
       "      <td>CD34</td>\n",
       "      <td>0.37037</td>\n",
       "      <td>ÂµM</td>\n",
       "      <td>500 nM</td>\n",
       "      <td>24</td>\n",
       "      <td>h</td>\n",
       "      <td>24 h</td>\n",
       "      <td>AML001_CD34_24H_X1_F1B10:J04|AML001_CD34_24H_X...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             pert_id     pert_iname  \\\n",
       "sig_id                                                                \n",
       "AML001_CD34_24H:A05                             DMSO           DMSO   \n",
       "AML001_CD34_24H:A06                             DMSO           DMSO   \n",
       "AML001_CD34_24H:B05                             DMSO           DMSO   \n",
       "AML001_CD34_24H:B06                             DMSO           DMSO   \n",
       "AML001_CD34_24H:BRD-A03772856:0.37037  BRD-A03772856  BRD-A03772856   \n",
       "\n",
       "                                         pert_type cell_id pert_dose  \\\n",
       "sig_id                                                                 \n",
       "AML001_CD34_24H:A05                    ctl_vehicle    CD34       0.1   \n",
       "AML001_CD34_24H:A06                    ctl_vehicle    CD34       0.1   \n",
       "AML001_CD34_24H:B05                    ctl_vehicle    CD34       0.1   \n",
       "AML001_CD34_24H:B06                    ctl_vehicle    CD34       0.1   \n",
       "AML001_CD34_24H:BRD-A03772856:0.37037       trt_cp    CD34   0.37037   \n",
       "\n",
       "                                      pert_dose_unit pert_idose  pert_time  \\\n",
       "sig_id                                                                       \n",
       "AML001_CD34_24H:A05                                %      0.1 %         24   \n",
       "AML001_CD34_24H:A06                                %      0.1 %         24   \n",
       "AML001_CD34_24H:B05                                %      0.1 %         24   \n",
       "AML001_CD34_24H:B06                                %      0.1 %         24   \n",
       "AML001_CD34_24H:BRD-A03772856:0.37037             ÂµM     500 nM         24   \n",
       "\n",
       "                                      pert_time_unit pert_itime  \\\n",
       "sig_id                                                            \n",
       "AML001_CD34_24H:A05                                h       24 h   \n",
       "AML001_CD34_24H:A06                                h       24 h   \n",
       "AML001_CD34_24H:B05                                h       24 h   \n",
       "AML001_CD34_24H:B06                                h       24 h   \n",
       "AML001_CD34_24H:BRD-A03772856:0.37037              h       24 h   \n",
       "\n",
       "                                                                               distil_id  \n",
       "sig_id                                                                                    \n",
       "AML001_CD34_24H:A05                                         AML001_CD34_24H_X1_F1B10:A05  \n",
       "AML001_CD34_24H:A06                                         AML001_CD34_24H_X3_F1B10:A06  \n",
       "AML001_CD34_24H:B05                    AML001_CD34_24H_X1_F1B10:B05|AML001_CD34_24H_X...  \n",
       "AML001_CD34_24H:B06                                         AML001_CD34_24H_X3_F1B10:B06  \n",
       "AML001_CD34_24H:BRD-A03772856:0.37037  AML001_CD34_24H_X1_F1B10:J04|AML001_CD34_24H_X...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#this files describe the perturbation metadata.\n",
    "#sig_id (index) is unique to a given measurement\n",
    "sig_info_gse92742=pd.read_csv('../data/GSE92742/GSE92742_Broad_LINCS_sig_info.txt',\n",
    "                             sep='\\t',header=0,index_col=0,low_memory=False)\n",
    "sig_info_gse92742.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pert_id</th>\n",
       "      <th>pert_iname</th>\n",
       "      <th>pert_type</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>pert_dose</th>\n",
       "      <th>pert_dose_unit</th>\n",
       "      <th>pert_idose</th>\n",
       "      <th>pert_time</th>\n",
       "      <th>pert_time_unit</th>\n",
       "      <th>pert_itime</th>\n",
       "      <th>distil_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sig_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CGS001_A375_96H:A2M:1</th>\n",
       "      <td>CGS001-2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>trt_sh.cgs</td>\n",
       "      <td>A375</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ÂµL</td>\n",
       "      <td>1 ÂµL</td>\n",
       "      <td>96</td>\n",
       "      <td>h</td>\n",
       "      <td>96 h</td>\n",
       "      <td>KDC004_A375_96H:TRCN0000006653:-666|KDC004_A37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_A375_96H:AARS:1</th>\n",
       "      <td>CGS001-16</td>\n",
       "      <td>AARS</td>\n",
       "      <td>trt_sh.cgs</td>\n",
       "      <td>A375</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ÂµL</td>\n",
       "      <td>1 ÂµL</td>\n",
       "      <td>96</td>\n",
       "      <td>h</td>\n",
       "      <td>96 h</td>\n",
       "      <td>KDB006_A375_96H:TRCN0000045688:-666|KDB006_A37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_A375_96H:AATF:1</th>\n",
       "      <td>CGS001-26574</td>\n",
       "      <td>AATF</td>\n",
       "      <td>trt_sh.cgs</td>\n",
       "      <td>A375</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ÂµL</td>\n",
       "      <td>1 ÂµL</td>\n",
       "      <td>96</td>\n",
       "      <td>h</td>\n",
       "      <td>96 h</td>\n",
       "      <td>KDA004_A375_96H:TRCN0000017389:-666|KDA004_A37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_A375_96H:ABAT:1</th>\n",
       "      <td>CGS001-18</td>\n",
       "      <td>ABAT</td>\n",
       "      <td>trt_sh.cgs</td>\n",
       "      <td>A375</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ÂµL</td>\n",
       "      <td>1 ÂµL</td>\n",
       "      <td>96</td>\n",
       "      <td>h</td>\n",
       "      <td>96 h</td>\n",
       "      <td>KDA002_A375_96H:TRCN0000034925:-666|KDB006_A37...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_A375_96H:ABCA1:1</th>\n",
       "      <td>CGS001-19</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>trt_sh.cgs</td>\n",
       "      <td>A375</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ÂµL</td>\n",
       "      <td>1 ÂµL</td>\n",
       "      <td>96</td>\n",
       "      <td>h</td>\n",
       "      <td>96 h</td>\n",
       "      <td>KDC004_A375_96H:TRCN0000029089:-666|KDC004_A37...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              pert_id pert_iname   pert_type cell_id  \\\n",
       "sig_id                                                                 \n",
       "CGS001_A375_96H:A2M:1        CGS001-2        A2M  trt_sh.cgs    A375   \n",
       "CGS001_A375_96H:AARS:1      CGS001-16       AARS  trt_sh.cgs    A375   \n",
       "CGS001_A375_96H:AATF:1   CGS001-26574       AATF  trt_sh.cgs    A375   \n",
       "CGS001_A375_96H:ABAT:1      CGS001-18       ABAT  trt_sh.cgs    A375   \n",
       "CGS001_A375_96H:ABCA1:1     CGS001-19      ABCA1  trt_sh.cgs    A375   \n",
       "\n",
       "                        pert_dose pert_dose_unit pert_idose  pert_time  \\\n",
       "sig_id                                                                   \n",
       "CGS001_A375_96H:A2M:1         1.0             ÂµL       1 ÂµL         96   \n",
       "CGS001_A375_96H:AARS:1        1.0             ÂµL       1 ÂµL         96   \n",
       "CGS001_A375_96H:AATF:1        1.0             ÂµL       1 ÂµL         96   \n",
       "CGS001_A375_96H:ABAT:1        1.0             ÂµL       1 ÂµL         96   \n",
       "CGS001_A375_96H:ABCA1:1       1.0             ÂµL       1 ÂµL         96   \n",
       "\n",
       "                        pert_time_unit pert_itime  \\\n",
       "sig_id                                              \n",
       "CGS001_A375_96H:A2M:1                h       96 h   \n",
       "CGS001_A375_96H:AARS:1               h       96 h   \n",
       "CGS001_A375_96H:AATF:1               h       96 h   \n",
       "CGS001_A375_96H:ABAT:1               h       96 h   \n",
       "CGS001_A375_96H:ABCA1:1              h       96 h   \n",
       "\n",
       "                                                                 distil_id  \n",
       "sig_id                                                                      \n",
       "CGS001_A375_96H:A2M:1    KDC004_A375_96H:TRCN0000006653:-666|KDC004_A37...  \n",
       "CGS001_A375_96H:AARS:1   KDB006_A375_96H:TRCN0000045688:-666|KDB006_A37...  \n",
       "CGS001_A375_96H:AATF:1   KDA004_A375_96H:TRCN0000017389:-666|KDA004_A37...  \n",
       "CGS001_A375_96H:ABAT:1   KDA002_A375_96H:TRCN0000034925:-666|KDB006_A37...  \n",
       "CGS001_A375_96H:ABCA1:1  KDC004_A375_96H:TRCN0000029089:-666|KDC004_A37...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#we select measurements where shRNA was used to knock-down a given gene\n",
    "#we will focuse singatures where pert_type column is a trt_sh.cgs, thw consensus signature of gene KD\n",
    "fil=sig_info_gse92742['pert_type']=='trt_sh.cgs'\n",
    "sig_info_shRNA=sig_info_gse92742[fil].copy()\n",
    "sig_info_shRNA.head()\n",
    "#here pert_iname is the knock-downed gene, cell_id is the cell line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pr_gene_symbol</th>\n",
       "      <th>pr_gene_title</th>\n",
       "      <th>pr_is_lm</th>\n",
       "      <th>pr_is_bing</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pr_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>780</th>\n",
       "      <td>DDR1</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7849</th>\n",
       "      <td>PAX8</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2978</th>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>guanylate cyclase activator 1A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2049</th>\n",
       "      <td>EPHB3</td>\n",
       "      <td>EPH receptor B3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2101</th>\n",
       "      <td>ESRRA</td>\n",
       "      <td>estrogen related receptor alpha</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pr_gene_symbol                                pr_gene_title  \\\n",
       "pr_gene_id                                                               \n",
       "780                  DDR1  discoidin domain receptor tyrosine kinase 1   \n",
       "7849                 PAX8                                 paired box 8   \n",
       "2978               GUCA1A               guanylate cyclase activator 1A   \n",
       "2049                EPHB3                              EPH receptor B3   \n",
       "2101                ESRRA              estrogen related receptor alpha   \n",
       "\n",
       "            pr_is_lm  pr_is_bing  \n",
       "pr_gene_id                        \n",
       "780                1           1  \n",
       "7849               1           1  \n",
       "2978               0           0  \n",
       "2049               0           1  \n",
       "2101               0           1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#we will use only the measured, landmark genes (for simplicity, and we also thrust measured genes more then infered)\n",
    "gene_ids=pd.read_csv('../data/GSE92742/GSE92742_Broad_LINCS_gene_info.txt',\n",
    "                    sep='\\t',header=0,index_col=0)\n",
    "gene_ids.head()\n",
    "#pr_gene_id is the gene id used by LINCS,pr_gene_symbol is the 'human readable' gene symbol\n",
    "#pr_is_lm is 1 for measured genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pr_gene_id\n",
       "780      DDR1\n",
       "7849     PAX8\n",
       "6193     RPS5\n",
       "23      ABCF1\n",
       "9552    SPAG7\n",
       "Name: pr_gene_symbol, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fil=gene_ids['pr_is_lm']==1\n",
    "gene_ids=gene_ids[fil]\n",
    "gene_ids=gene_ids['pr_gene_symbol']\n",
    "gene_ids.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>rid</th>\n",
       "      <th>780</th>\n",
       "      <th>7849</th>\n",
       "      <th>6193</th>\n",
       "      <th>23</th>\n",
       "      <th>9552</th>\n",
       "      <th>387</th>\n",
       "      <th>10921</th>\n",
       "      <th>10285</th>\n",
       "      <th>533</th>\n",
       "      <th>6194</th>\n",
       "      <th>...</th>\n",
       "      <th>54681</th>\n",
       "      <th>11000</th>\n",
       "      <th>6915</th>\n",
       "      <th>6253</th>\n",
       "      <th>7264</th>\n",
       "      <th>5467</th>\n",
       "      <th>2767</th>\n",
       "      <th>23038</th>\n",
       "      <th>57048</th>\n",
       "      <th>79716</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CGS001_A375_96H:A2M:1</th>\n",
       "      <td>-0.450660</td>\n",
       "      <td>0.613190</td>\n",
       "      <td>-0.052799</td>\n",
       "      <td>-0.690646</td>\n",
       "      <td>-0.884775</td>\n",
       "      <td>-0.268257</td>\n",
       "      <td>0.989756</td>\n",
       "      <td>-0.270924</td>\n",
       "      <td>0.562012</td>\n",
       "      <td>-0.064880</td>\n",
       "      <td>...</td>\n",
       "      <td>0.769396</td>\n",
       "      <td>-0.574408</td>\n",
       "      <td>0.037353</td>\n",
       "      <td>-0.094501</td>\n",
       "      <td>0.246233</td>\n",
       "      <td>0.055728</td>\n",
       "      <td>-0.032736</td>\n",
       "      <td>0.554488</td>\n",
       "      <td>0.042339</td>\n",
       "      <td>-0.875408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_A549_96H:A2M:1</th>\n",
       "      <td>-1.007425</td>\n",
       "      <td>-0.465669</td>\n",
       "      <td>-0.152609</td>\n",
       "      <td>-0.854945</td>\n",
       "      <td>-0.173800</td>\n",
       "      <td>0.701383</td>\n",
       "      <td>0.440690</td>\n",
       "      <td>0.388689</td>\n",
       "      <td>0.346691</td>\n",
       "      <td>0.001048</td>\n",
       "      <td>...</td>\n",
       "      <td>0.821364</td>\n",
       "      <td>-1.732306</td>\n",
       "      <td>-0.990096</td>\n",
       "      <td>-2.820324</td>\n",
       "      <td>0.476273</td>\n",
       "      <td>-0.699089</td>\n",
       "      <td>0.398378</td>\n",
       "      <td>0.025279</td>\n",
       "      <td>0.423876</td>\n",
       "      <td>-1.017771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_HA1E_96H:A2M:1.5</th>\n",
       "      <td>-0.487499</td>\n",
       "      <td>0.373099</td>\n",
       "      <td>0.906173</td>\n",
       "      <td>-0.472660</td>\n",
       "      <td>-0.973289</td>\n",
       "      <td>1.075552</td>\n",
       "      <td>-0.182322</td>\n",
       "      <td>-0.518891</td>\n",
       "      <td>0.030982</td>\n",
       "      <td>-0.204590</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.289508</td>\n",
       "      <td>0.623188</td>\n",
       "      <td>-1.103047</td>\n",
       "      <td>-0.815189</td>\n",
       "      <td>0.651688</td>\n",
       "      <td>-0.208761</td>\n",
       "      <td>0.345416</td>\n",
       "      <td>0.606831</td>\n",
       "      <td>1.363954</td>\n",
       "      <td>-1.080051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_HEPG2_96H:A2M:1.5</th>\n",
       "      <td>-0.577616</td>\n",
       "      <td>-0.594645</td>\n",
       "      <td>0.088793</td>\n",
       "      <td>-0.679613</td>\n",
       "      <td>-1.105863</td>\n",
       "      <td>0.353294</td>\n",
       "      <td>0.151503</td>\n",
       "      <td>-0.299633</td>\n",
       "      <td>-0.678864</td>\n",
       "      <td>-0.074211</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.222315</td>\n",
       "      <td>-0.324444</td>\n",
       "      <td>-0.194537</td>\n",
       "      <td>-1.031143</td>\n",
       "      <td>0.009928</td>\n",
       "      <td>-0.348788</td>\n",
       "      <td>0.813288</td>\n",
       "      <td>1.007242</td>\n",
       "      <td>1.392832</td>\n",
       "      <td>0.153785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGS001_HT29_96H:A2M:1</th>\n",
       "      <td>-0.272184</td>\n",
       "      <td>-0.172414</td>\n",
       "      <td>0.200608</td>\n",
       "      <td>-0.235290</td>\n",
       "      <td>-0.262455</td>\n",
       "      <td>0.432267</td>\n",
       "      <td>-0.113109</td>\n",
       "      <td>-0.146903</td>\n",
       "      <td>-0.888957</td>\n",
       "      <td>0.007347</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.056524</td>\n",
       "      <td>-0.510606</td>\n",
       "      <td>-0.666978</td>\n",
       "      <td>-0.377540</td>\n",
       "      <td>0.178874</td>\n",
       "      <td>0.180162</td>\n",
       "      <td>1.115406</td>\n",
       "      <td>0.154438</td>\n",
       "      <td>0.346749</td>\n",
       "      <td>-0.103932</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 978 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "rid                            780      7849      6193        23      9552  \\\n",
       "cid                                                                          \n",
       "CGS001_A375_96H:A2M:1    -0.450660  0.613190 -0.052799 -0.690646 -0.884775   \n",
       "CGS001_A549_96H:A2M:1    -1.007425 -0.465669 -0.152609 -0.854945 -0.173800   \n",
       "CGS001_HA1E_96H:A2M:1.5  -0.487499  0.373099  0.906173 -0.472660 -0.973289   \n",
       "CGS001_HEPG2_96H:A2M:1.5 -0.577616 -0.594645  0.088793 -0.679613 -1.105863   \n",
       "CGS001_HT29_96H:A2M:1    -0.272184 -0.172414  0.200608 -0.235290 -0.262455   \n",
       "\n",
       "rid                            387     10921     10285       533      6194  \\\n",
       "cid                                                                          \n",
       "CGS001_A375_96H:A2M:1    -0.268257  0.989756 -0.270924  0.562012 -0.064880   \n",
       "CGS001_A549_96H:A2M:1     0.701383  0.440690  0.388689  0.346691  0.001048   \n",
       "CGS001_HA1E_96H:A2M:1.5   1.075552 -0.182322 -0.518891  0.030982 -0.204590   \n",
       "CGS001_HEPG2_96H:A2M:1.5  0.353294  0.151503 -0.299633 -0.678864 -0.074211   \n",
       "CGS001_HT29_96H:A2M:1     0.432267 -0.113109 -0.146903 -0.888957  0.007347   \n",
       "\n",
       "rid                       ...     54681     11000      6915      6253  \\\n",
       "cid                       ...                                           \n",
       "CGS001_A375_96H:A2M:1     ...  0.769396 -0.574408  0.037353 -0.094501   \n",
       "CGS001_A549_96H:A2M:1     ...  0.821364 -1.732306 -0.990096 -2.820324   \n",
       "CGS001_HA1E_96H:A2M:1.5   ... -0.289508  0.623188 -1.103047 -0.815189   \n",
       "CGS001_HEPG2_96H:A2M:1.5  ... -0.222315 -0.324444 -0.194537 -1.031143   \n",
       "CGS001_HT29_96H:A2M:1     ... -0.056524 -0.510606 -0.666978 -0.377540   \n",
       "\n",
       "rid                           7264      5467      2767     23038     57048  \\\n",
       "cid                                                                          \n",
       "CGS001_A375_96H:A2M:1     0.246233  0.055728 -0.032736  0.554488  0.042339   \n",
       "CGS001_A549_96H:A2M:1     0.476273 -0.699089  0.398378  0.025279  0.423876   \n",
       "CGS001_HA1E_96H:A2M:1.5   0.651688 -0.208761  0.345416  0.606831  1.363954   \n",
       "CGS001_HEPG2_96H:A2M:1.5  0.009928 -0.348788  0.813288  1.007242  1.392832   \n",
       "CGS001_HT29_96H:A2M:1     0.178874  0.180162  1.115406  0.154438  0.346749   \n",
       "\n",
       "rid                          79716  \n",
       "cid                                 \n",
       "CGS001_A375_96H:A2M:1    -0.875408  \n",
       "CGS001_A549_96H:A2M:1    -1.017771  \n",
       "CGS001_HA1E_96H:A2M:1.5  -1.080051  \n",
       "CGS001_HEPG2_96H:A2M:1.5  0.153785  \n",
       "CGS001_HT29_96H:A2M:1    -0.103932  \n",
       "\n",
       "[5 rows x 978 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#an example to select perturbations where A2M gene was knocked-down\n",
    "fil=sig_info_shRNA['pert_iname']=='A2M'\n",
    "samples=sig_info_shRNA.index[fil]\n",
    "#cid is the sample ides, while rid is the gene ids\n",
    "try:\n",
    "    gene_ids.index=gene_ids.index.astype(str) #in some version gene ids are string\n",
    "    expression=parse('../data/GSE92742/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx',\n",
    "                 cid=samples,rid=gene_ids.index).data_df.T[gene_ids.index]\n",
    "except:\n",
    "    gene_ids.index=gene_ids.index.astype(int) # in others integer\n",
    "    expression=parse('../data/GSE92742/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx',\n",
    "                 cid=samples,rid=gene_ids.index).data_df.T[gene_ids.index]\n",
    "expression.head()\n",
    "#so here the rows are the samples where the given genes were knocked down\n",
    "#columsn are the measured gene expression changes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "In this analysis we won't work with the individual signatures, but will caculate a consensus (average) signature for a given gene knockout. However further filtering / using of individual signatures are highly encouraged. To calculate consensus signature, we will use the MODZ method described in the original LINCS manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "from scipy.stats import spearmanr as scor\n",
    "def calc_MODZ(data):\n",
    "    \"\"\"calculates MODZ based on the original CMAP/L1000 study\n",
    "    use only lm genes for MODZ calculation! Uses LM_GENES global\n",
    "    variable.\"\"\"\n",
    "    if len(data)==1:\n",
    "        return data.iloc[0]\n",
    "    if len(data)==2:\n",
    "        return np.mean(data,0)\n",
    "    else:\n",
    "        CM=scor(data.T)[0]\n",
    "        fil=CM<0\n",
    "        CM[fil]=0.01\n",
    "        weights=np.sum(CM,1)-1\n",
    "        weights=weights/np.sum(weights)\n",
    "        weights=weights.reshape((-1,1))\n",
    "        return pd.Series(np.dot(data.T,weights).reshape((-1,1)[0]),index=data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "rid\n",
       "780    -0.370345\n",
       "7849   -0.021135\n",
       "6193    0.446688\n",
       "23     -0.631813\n",
       "9552   -0.688607\n",
       "dtype: float64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#so this is the consensus signature of the A2M gene knockdown\n",
    "calc_MODZ(expression).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>pr_gene_id</th>\n",
       "      <th>780</th>\n",
       "      <th>7849</th>\n",
       "      <th>6193</th>\n",
       "      <th>23</th>\n",
       "      <th>9552</th>\n",
       "      <th>387</th>\n",
       "      <th>10921</th>\n",
       "      <th>10285</th>\n",
       "      <th>533</th>\n",
       "      <th>6194</th>\n",
       "      <th>...</th>\n",
       "      <th>54681</th>\n",
       "      <th>11000</th>\n",
       "      <th>6915</th>\n",
       "      <th>6253</th>\n",
       "      <th>7264</th>\n",
       "      <th>5467</th>\n",
       "      <th>2767</th>\n",
       "      <th>23038</th>\n",
       "      <th>57048</th>\n",
       "      <th>79716</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARS</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AATF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABAT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCA1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 978 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "pr_gene_id  780 7849 6193   23 9552  387 10921 10285  533 6194  ... 54681  \\\n",
       "A2M         NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN  NaN  NaN  ...   NaN   \n",
       "AARS        NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN  NaN  NaN  ...   NaN   \n",
       "AATF        NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN  NaN  NaN  ...   NaN   \n",
       "ABAT        NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN  NaN  NaN  ...   NaN   \n",
       "ABCA1       NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN  NaN  NaN  ...   NaN   \n",
       "\n",
       "pr_gene_id 11000 6915 6253 7264 5467 2767 23038 57048 79716  \n",
       "A2M          NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   NaN  \n",
       "AARS         NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   NaN  \n",
       "AATF         NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   NaN  \n",
       "ABAT         NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   NaN  \n",
       "ABCA1        NaN  NaN  NaN  NaN  NaN  NaN   NaN   NaN   NaN  \n",
       "\n",
       "[5 rows x 978 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#we will store perturabtion expression profiles in a DataFrame\n",
    "#rows are the perturbed genes, columns are the measured landmark genes from LINCS\n",
    "genes_perturbed=sig_info_shRNA['pert_iname'].unique()\n",
    "consensus_signatures_shRNA=pd.DataFrame(index=genes_perturbed,columns=gene_ids.index)\n",
    "consensus_signatures_shRNA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done for 0 genes\n",
      "Done for 100 genes\n",
      "Done for 200 genes\n",
      "Done for 300 genes\n",
      "Done for 400 genes\n",
      "Done for 500 genes\n",
      "Done for 600 genes\n",
      "Done for 700 genes\n",
      "Done for 800 genes\n",
      "Done for 900 genes\n",
      "Done for 1000 genes\n",
      "Done for 1100 genes\n",
      "Done for 1200 genes\n",
      "Done for 1300 genes\n",
      "Done for 1400 genes\n",
      "Done for 1500 genes\n",
      "Done for 1600 genes\n",
      "Done for 1700 genes\n",
      "Done for 1800 genes\n",
      "Done for 1900 genes\n",
      "Done for 2000 genes\n",
      "Done for 2100 genes\n",
      "Done for 2200 genes\n",
      "Done for 2300 genes\n",
      "Done for 2400 genes\n",
      "Done for 2500 genes\n",
      "Done for 2600 genes\n",
      "Done for 2700 genes\n",
      "Done for 2800 genes\n",
      "Done for 2900 genes\n",
      "Done for 3000 genes\n",
      "Done for 3100 genes\n",
      "Done for 3200 genes\n",
      "Done for 3300 genes\n",
      "Done for 3400 genes\n",
      "Done for 3500 genes\n",
      "Done for 3600 genes\n",
      "Done for 3700 genes\n",
      "Done for 3800 genes\n",
      "Done for 3900 genes\n",
      "Done for 4000 genes\n",
      "Done for 4100 genes\n",
      "Done for 4200 genes\n",
      "Done for 4300 genes\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(genes_perturbed)):\n",
    "    if (i%100)==0:\n",
    "        print('Done for %i genes' %i)\n",
    "    gene=genes_perturbed[i]\n",
    "    fil=sig_info_shRNA['pert_iname']==gene\n",
    "    samples=sig_info_shRNA.index[fil]\n",
    "    expression=parse('../data/GSE92742/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx',\n",
    "                 cid=samples,rid=gene_ids.index).data_df.T[gene_ids.index]\n",
    "    consensus_signatures_shRNA.loc[gene]=calc_MODZ(expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#rename LINCS gene ids for gene symbolc\n",
    "consensus_signatures_shRNA.columns=gene_ids[consensus_signatures_shRNA.columns].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "consensus_signatures_shRNA.to_csv('../results/consensus_signature_shRNA.csv',sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "So you can use the **consensus_signature_shRNA.csv** file for your models. The same file is also available in Synapse. While this data only contains consensus signatures, further experiments with individual signatures can further increase model perferomance, and are highly encouraged."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Now we will do the same consensus signature calculations for drugs. We will use both GSE92742 and GSE70138."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "sig_info_gse92742=pd.read_csv('../data/GSE92742/GSE92742_Broad_LINCS_sig_info.txt',\n",
    "                              sep='\\t',header=0,index_col=0,low_memory=False)\n",
    "sig_info_gse70138=pd.read_csv('../data/GSE70138/GSE70138_Broad_LINCS_sig_info.txt',\n",
    "                              sep='\\t',header=0,index_col=0,low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#keep only compoud treatments\n",
    "fil=sig_info_gse92742['pert_type']=='trt_cp'\n",
    "sig_info_gse92742=sig_info_gse92742[fil]\n",
    "fil=sig_info_gse70138['pert_type']=='trt_cp'\n",
    "sig_info_gse70138=sig_info_gse70138[fil]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#get all compounds\n",
    "cpds=list(set(sig_info_gse70138['pert_id']) | set(sig_info_gse92742['pert_id']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "To get the targets of these drugs - when target is known - we need compound metadata from Drug Repurposing Hub."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# read drug repurposing data\n",
    "repurposing=pd.read_csv('../data/repurposing/repurposing_drugs_20180907.txt',sep='\\t',encoding='latin',\n",
    "                    header=0,index_col=None,skiprows=9)\n",
    "repurposing_samples=pd.read_csv('../data/repurposing/repurposing_samples_20180907.txt',sep='\\t',encoding='latin',\n",
    "                    header=0,index_col=None,skiprows=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#correct broad_id\n",
    "repurposing_samples['broad_id']=repurposing_samples['broad_id'].apply(lambda x: '-'.join(x.split('-')[:2]))\n",
    "#select usefull columns\n",
    "repurposing_samples=repurposing_samples[['broad_id','pert_iname']]\n",
    "repurposing=repurposing[['pert_iname','moa','target']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#merge repurposing data\n",
    "drug_metadata=pd.merge(repurposing,repurposing_samples)\n",
    "#remove duplicated\n",
    "drug_metadata=drug_metadata.drop_duplicates()\n",
    "#keep only rows with known target / MoA\n",
    "fil=~(drug_metadata['moa'].isnull() & drug_metadata['target'].isnull())\n",
    "drug_metadata=drug_metadata[fil]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#matching compounds between repurposing and LINCS dataset\n",
    "cpds=list(set(cpds) & set(drug_metadata['broad_id']))\n",
    "fil=np.in1d(drug_metadata['broad_id'],cpds)\n",
    "drug_metadata=drug_metadata[fil]\n",
    "drug_metadata.index=range(len(drug_metadata))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#select all targets / MoAs\n",
    "all_moas=[]\n",
    "all_targets=[]\n",
    "for i in drug_metadata.index:\n",
    "    if not pd.isnull(drug_metadata.loc[i,'target']):\n",
    "        all_targets+=drug_metadata.loc[i,'target'].split('|')\n",
    "    if  not pd.isnull(drug_metadata.loc[i,'moa']):\n",
    "        all_moas+=drug_metadata.loc[i,'moa'].split('|')\n",
    "all_moas=list(set(all_moas))\n",
    "all_targets=list(set(all_targets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "meta_matrix=pd.DataFrame(0,index=list(set(drug_metadata['broad_id'])),columns=all_moas+all_targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done for 0 drugs\n",
      "Done for 100 drugs\n",
      "Done for 200 drugs\n",
      "Done for 300 drugs\n",
      "Done for 400 drugs\n",
      "Done for 500 drugs\n",
      "Done for 600 drugs\n",
      "Done for 700 drugs\n",
      "Done for 800 drugs\n",
      "Done for 900 drugs\n",
      "Done for 1000 drugs\n",
      "Done for 1100 drugs\n",
      "Done for 1200 drugs\n",
      "Done for 1300 drugs\n",
      "Done for 1400 drugs\n",
      "Done for 1500 drugs\n",
      "Done for 1600 drugs\n",
      "Done for 1700 drugs\n",
      "Done for 1800 drugs\n",
      "Done for 1900 drugs\n",
      "Done for 2000 drugs\n",
      "Done for 2100 drugs\n",
      "Done for 2200 drugs\n",
      "Done for 2300 drugs\n",
      "Done for 2400 drugs\n",
      "Done for 2500 drugs\n",
      "Done for 2600 drugs\n",
      "Done for 2700 drugs\n",
      "Done for 2800 drugs\n"
     ]
    }
   ],
   "source": [
    "#for compunds where inhibitor,blocker,antagonist etc. are in the MoA columns\n",
    "#we assume they are inhibitory compounds, so we will mark them with -1 in the meta_matrix\n",
    "# for other compounds, we assume they are activators, we will mark them with +1\n",
    "#this is probably not a perfect way to access inhibitory/acovatory state, but good for a first try\n",
    "inhibitory_words=set(['inhibitor','blocker','antagonist','inihibitor']) #inihibitor is just a typo\n",
    "for i in drug_metadata.index:\n",
    "    if list(drug_metadata.index).index(i) % 100==0:\n",
    "        print('Done for %i drugs' %list(drug_metadata.index).index(i))\n",
    "    brd=drug_metadata.loc[i,'broad_id']\n",
    "    if not pd.isnull(drug_metadata.loc[i,'moa']):\n",
    "        moas=drug_metadata.loc[i,'moa'].split('|')\n",
    "    else:\n",
    "        moas=[]\n",
    "    if not pd.isnull(drug_metadata.loc[i,'target']):\n",
    "        s=1\n",
    "        targets=drug_metadata.loc[i,'target'].split('|')\n",
    "        if len(set((' '.join(moas)).split())&inhibitory_words)>0:\n",
    "            s=-1\n",
    "    else:\n",
    "        targets=[]\n",
    "    meta_matrix.loc[brd,moas]=1\n",
    "    meta_matrix.loc[brd,targets]=s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "meta_matrix.to_csv('../results/drugs_meta.csv',sep=',')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Now we will get the consensus signature for each drug, and store them in a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "consensus_signatures_drugs=pd.DataFrame(index=meta_matrix.index,columns=gene_ids.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done for 0 drugs\n",
      "Done for 100 drugs\n",
      "Done for 200 drugs\n",
      "Done for 300 drugs\n",
      "Done for 400 drugs\n",
      "Done for 500 drugs\n",
      "Done for 600 drugs\n",
      "Done for 700 drugs\n",
      "Done for 800 drugs\n",
      "Done for 900 drugs\n",
      "Done for 1000 drugs\n",
      "Done for 1100 drugs\n",
      "Done for 1200 drugs\n",
      "Done for 1300 drugs\n",
      "Done for 1400 drugs\n",
      "Done for 1500 drugs\n",
      "Done for 1600 drugs\n",
      "Done for 1700 drugs\n",
      "Done for 1800 drugs\n",
      "Done for 1900 drugs\n",
      "Done for 2000 drugs\n",
      "Done for 2100 drugs\n",
      "Done for 2200 drugs\n",
      "Done for 2300 drugs\n",
      "Done for 2400 drugs\n",
      "Done for 2500 drugs\n",
      "Done for 2600 drugs\n",
      "Done for 2700 drugs\n",
      "Done for 2800 drugs\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(consensus_signatures_drugs.index)):\n",
    "    drug=consensus_signatures_drugs.index[i]\n",
    "    if i%100==0:\n",
    "        print('Done for %i drugs' % i)\n",
    "    fil=sig_info_gse70138['pert_id']==drug\n",
    "    if fil.sum()>0:\n",
    "        samples=sig_info_gse70138.index[fil]\n",
    "        expression_gse70138=parse('../data/GSE70138/GSE70138_Broad_LINCS_Level5_COMPZ_n118050x12328.gctx',\n",
    "                                  cid=samples,rid=gene_ids.index).data_df.T[gene_ids.index]\n",
    "    else:\n",
    "        expression_gse70138=pd.DataFrame(columns=gene_ids.index)\n",
    "    fil=sig_info_gse92742['pert_id']==drug\n",
    "    if fil.sum()>0:\n",
    "        samples=sig_info_gse92742.index[fil]\n",
    "        expression_gse92742=parse('../data/GSE92742/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx',\n",
    "                                  cid=samples,rid=gene_ids.index).data_df.T[gene_ids.index]\n",
    "    else:\n",
    "        expression_gse92742=pd.DataFrame(columns=gene_ids.index)\n",
    "    expression=pd.concat([expression_gse70138,expression_gse92742])\n",
    "    consensus_signatures_drugs.loc[drug]=calc_MODZ(expression)\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#rename L1000 gene ids for gene names\n",
    "consensus_signatures_drugs.columns=gene_ids[consensus_signatures_drugs.columns].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "consensus_signatures_drugs.to_csv('../results/consensus_signature_drugs.csv',sep=',')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
